posts/Gunnar/2017-05-29-creating-patches-for-wildfly.adoc

```diff
@@ -0,0 +1,361 @@
+    </resources>
+
+    <dependencies>
+        <module name="org.eclipse.collections.api" />
+    </dependencies>
+</module>
+----
+
+Each descriptor specifies a resource for the corresponding JAR
+(the version property placeholders are replaced using https://maven.apache.org/plugins/maven-resources-plugin/examples/filter.html[Maven resource filtering]).
+The implementation module also declares a dependency to the API module.
+
+== The Patch Tool Configuration File
+
+The _patch-gen_ tool needs a small configuration file which describes some patch metadata
+(e.g. the server version to which the patch applies and the type of the patch - one-off vs. cumulative) as well as the patched module(s):
+
+[source,java]
+._src/main/patch/patch.xml_
+----
+<?xml version='1.0' encoding='UTF-8'?>
+
+<patch-config xmlns="urn:jboss:patch-config:1.0">
+    <name>wildfly-${wildfly.version}-eclipse-collections-${eclipse.collections.version}</name>
+    <description>This patch adds Eclipse Collections ${eclipse.collections.version} to a WildFly ${wildfly.version} installation</description>
+    <element patch-id="layer-base-wildfly-${wildfly.version}-eclipse-collections-${eclipse.collections.version}">
+        <one-off name="base" />
+        <description>This patch adds Eclipse Collections ${eclipse.collections.version} to a WildFly ${wildfly.version} installation</description>
+        <specified-content>
+            <modules>
+                <added name="org.eclipse.collections.api" />
+                <added name="org.eclipse.collections" />
+            </modules>
+        </specified-content>
+    </element>
+    <specified-content/>
+</patch-config>
+----
+
+== Preparing the Patch Creation
+
+The _patch-gen_ tool takes two directories of the distribution to be patched as input:
+one directory with the original, unpatched WildFly structure and another directory which contains the original WildFly structure _and_ the added (or updated) modules.
+We can use the https://maven.apache.org/plugins/maven-dependency-plugin/[Maven dependency plug-in] for creating the two directories by extracting the WildFly distribution twice:
+
+[source,xml]
+._pom.xml_
+----
+...
+<plugin>
+    <artifactId>maven-dependency-plugin</artifactId>
+    <executions>
+        <execution>
+            <id>unpack-wildfly</id>
+            <phase>prepare-package</phase>
+            <goals>
+                <goal>unpack</goal>
+            </goals>
+            <configuration>
+                <artifactItems>
+                    <artifactItem>
+                        <groupId>org.wildfly</groupId>
+                        <artifactId>wildfly-dist</artifactId>
+                        <version>${wildfly.version}</version>
+                        <type>tar.gz</type>
+                        <overWrite>false</overWrite>
+                        <outputDirectory>${project.build.directory}/wildfly-original</outputDirectory>
```

Comment:
Those vars contain the version at the end: 

    <wildflyOriginal>${project.build.directory}/wildfly-original/wildfly-${wildfly.version}</wildflyOriginal>
    <wildflyPatched>${project.build.directory}/wildfly-patched/wildfly-${wildfly.version}</wildflyPatched>

I guess one could strip that top-level dir when unpacking somehow, but it doesn't seem worth the extra effort.